### PR TITLE
Remove distutils

### LINKED ISSUE #56527: Remove distutils
Distutils is [deprecated via PEP 632](https://www.python.org/dev/peps/pep-0632/#:~:text=In%20Python%203.10%20and%203.11,platforms%20will%20not%20be%20added.) in Python 3.10+ and will be removed in 3.12. There's no reason for us to keep using it, so we should get rid of all usages so it doesn't bite us later when 3.10 rolls around.

Usages:

```
$ ag --ignore third_party distutils
caffe2/CMakeLists.txt
1707:      from distutils import sysconfig
1719:      from distutils import sysconfig

cmake/Dependencies.cmake
938:    # distutils.sysconfig, if it's installed, is more accurate than sysconfig,
940:    pycmd_no_exit(_py_inc _exitcode "from distutils import sysconfig; print(sysconfig.get_python_inc())")
943:      message(STATUS "Setting Python's include dir to ${_py_inc} from distutils.sysconfig")

scripts/build_android.sh
64:  CMAKE_ARGS+=("-DCMAKE_PREFIX_PATH=$($PYTHON -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())')")

scripts/build_mobile.sh
18:CMAKE_ARGS+=("-DCMAKE_PREFIX_PATH=$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())')")

scripts/get_python_cmake_flags.py
18:from distutils import sysconfig

scripts/build_ios.sh
16:  CMAKE_ARGS+=("-DCMAKE_PREFIX_PATH=$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())')")

test/test_spectral_ops.py
16:from distutils.version import LooseVersion

tools/setup_helpers/cmake.py
10:import distutils.sysconfig
11:from distutils.version import LooseVersion
233:            'CMAKE_PREFIX_PATH': distutils.sysconfig.get_python_lib()
319:                      PYTHON_INCLUDE_DIR=distutils.sysconfig.get_python_inc(),

tools/build_pytorch_libs.py
10:    from distutils._msvccompiler import _get_vc_env

tools/generate_torch_version.py
5:from distutils.util import strtobool

torch/package/_stdlib.py
79:    "distutils",
298:    "distutils",
516:    "distutils",
732:    "distutils",

torch/testing/_internal/common_cuda.py
9:from distutils.version import LooseVersion

torch/testing/_internal/common_methods_invocations.py
37:from distutils.version import LooseVersion

torch/utils/tensorboard/__init__.py
2:from distutils.version import LooseVersion

torch/utils/cpp_extension.py
338:    Fallbacks to the standard distutils backend if Ninja is not available.
368:                   '{}. Falling back to using the slow distutils backend.')
479:            # of distutils.UnixCCompiler). See the following link.
480:            # https://github.com/python/cpython/blob/f03a8f8d5001963ad5b5b28dbd95497e9cc15596/Lib/distutils/ccompiler.py#L564-L567
484:            # we update our python version (which is when distutils can be
697:        # https://github.com/python/cpython/blob/dc0284ee8f7a270b6005467f26d8e5773d76e959/Lib/distutils/ccompiler.py#L511
737:        if IS_WINDOWS and 'VSCMD_ARG_TGT_ARCH' in os.environ and 'DISTUTILS_USE_SDK' not in os.environ:
738:            msg = ('It seems that the VC environment is activated but DISTUTILS_USE_SDK is not set.'
740:                   'Please set `DISTUTILS_USE_SDK=1` and try again.')
1638:        from distutils.util import get_platform
1639:        from distutils._msvccompiler import _get_vc_env

README.md
316:set DISTUTILS_USE_SDK=1

mypy.ini
190:[mypy-distutils.*]

setup.py
24:#     the C/C++ compiler to use (NB: the CXX flag has no effect for distutils
25:#     compiles, because distutils always uses CC to compile, even for C++
200:from distutils import core
201:from distutils.core import Distribution
202:from distutils.errors import DistutilsArgError
205:import distutils.command.clean
206:import distutils.command.sdist
207:import distutils.sysconfig
270:rel_site_packages = distutils.sysconfig.get_python_lib(prefix='')
272:full_site_packages = distutils.sysconfig.get_python_lib()
276:        distutils.sysconfig.get_config_var("prefix"),
277:        distutils.sysconfig.get_config_var("VERSION"))
283:            distutils.sysconfig.get_config_var("VERSION"))
286:        distutils.sysconfig.get_config_var("LIBDIR"),
287:        distutils.sysconfig.get_config_var("INSTSONAME"))
288:cmake_python_include_dir = distutils.sysconfig.get_python_inc()
494:        system_c_flags = str(distutils.sysconfig.get_config_var('CFLAGS'))
496:            os.environ['CC'] = str(distutils.sysconfig.get_config_var('CC'))
562:        distutils.command.build_ext.build_ext.build_extensions(self)
565:        outputs = distutils.command.build_ext.build_ext.get_outputs(self)
647:class clean(distutils.command.clean.clean):
671:class sdist(distutils.command.sdist.sdist):
870:    except DistutilsArgError as msg:
```

cc @malfet @seemethere @walterddr @pytorch/pytorch-dev-infra

### PR DESCRIPTION
**Pull request by driazati**
[distutils](https://docs.python.org/3/library/distutils.html) is on its way out and will be deprecated-on-import for Python 3.10+ and removed in Python 3.12 (see [PEP 632](https://www.python.org/dev/peps/pep-0632/)). There's no reason for us to keep it around since all the functionality we want from it can be found in `setuptools` / `sysconfig`.

Fixes #56527

Differential Revision: [D28051356](https://our.internmc.facebook.com/intern/diff/28051356/)

### CODE REVIEW

File: caffe2/CMakeLists.txt [hunk 1]
```diff
@@ -1697,8 +1697,9 @@ if(BUILD_PYTHON)
   # https://github.com/pytorch/pytorch/tree/master/tools/build_pytorch_libs.bat#note-backslash-munging-on-windows
   pycmd(PYTHON_SITE_PACKAGES "
       import os
-      from distutils import sysconfig
-      print(sysconfig.get_python_lib(prefix=''))
+      import sysconfig
+      relative_site_packages = sysconfig.get_path('purelib').replace(sysconfig.get_path('data'), '').lstrip(os.path.sep)
+      print(relative_site_packages)
```
Comment (ezyang) [inline caffe2/CMakeLists.txt:1702]: the heck is this

**Comment by ezyang** [comment]
For the most part things look simpler. But two spots look very suspect, both in `caffe2/CMakeLists.txt`, where the logic has gotten substantially longer. What's going on?

File: caffe2/CMakeLists.txt (hunk 1 above)
Comment (driazati) [reply caffe2/CMakeLists.txt:1702]: `sysconfig` [hardcodes these paths with templates](https://github.com/python/cpython/blob/103d5e420dd90489933ad9da8bb1d6008773384d/Lib/sysconfig.py#L27-L38), but `distutils.sysconfig` doesn't and lets you supply your own prefix (i.e. `base` in the link), so here we end up having to remove the `base` (and its trailing slash) prefix manually.

```
>>> from distutils import sysconfig

>>> sysconfig.get_python_lib()
'/home/davidriazati/local/miniconda3/envs/pytorch/lib/python3.9/site-packages'

>>> sysconfig.get_python_lib(prefix='')
'lib/python3.9/site-packages'
```


```
>>> import sysconfig

>>> sysconfig.get_path('purelib')
'/home/davidriazati/local/miniconda3/envs/pytorch/lib/python3.9/site-packages'

>>> sysconfig.get_path('data') 
'/home/davidriazati/local/miniconda3/envs/pytorch'

```

**Comment by driazati** [comment]
See comment about the first weird thing, for the second it's only to work around that bug linked in the comments (the fix for which is in 3.8+ but isn't getting backported any further than that). There are definitely some un-addressed gaps in the deprecation PEP but it's accepted so 🤷

**Review by ezyang — APPROVED** [review]
okey dokey

**Comment by xkszltl** [comment]
I noticed on Ubuntu 18.04, `dist-packages/caffe2` has been moved from `/usr/local/lib/python3` to `/usr/local/lib/python3.6` recently.
I love the move because it's finally consistent with other distros, but not sure which commit introduced the behavior, to check whether it is a reliable change.
So far I only found this PR to be suspicious.
IIRC Ubuntu has some special patches for distutils to create things like dist-packages vs. site-packages.
If so, could you have confirm?

Thanks in advance.

**Comment by xkszltl** [comment]
I can confirm the change is related to this PR:
```
# lsb_release -d
Description:    Ubuntu 18.04.5 LTS
# python3 -c 'import sysconfig, distutils.sysconfig; print(sysconfig.get_path("purelib"), distutils.sysconfig.get_python_lib())'
/usr/lib/python3.6/site-packages /usr/lib/python3/dist-packages
```

Another thing I want to mention is, seems by default Debian python only load from dist-packages:
```
# python3 -c 'import sys; print(sys.path)'
['', '/usr/lib/python36.zip', '/usr/lib/python3.6', '/usr/lib/python3.6/lib-dynload', '/usr/local/lib/python3.6/dist-packages', '/usr/lib/python3/dist-packages']
```

**Comment by ezyang** [comment]
@xkszltl Thanks for the diagnosis. Do you think there's anything we need to do?

**Comment by xkszltl** [comment]
> Do you think there's anything we need to do?

1. Confirm if this is an intentional change (so that it won't be "fixed"). If so I can start depending on the new result.
2. Double check if Debian/Ubuntu can pick it up from site-packages without extra `PYTHONPATH`.

**Comment by driazati** [comment]
This shouldn't affect wheel installs since those are all relative from the `.whl` file zip structure, but I think it would mess up builds from source on Debian/Ubuntu that don't use conda (i.e. just using the bare system Python). I'm not sure if we even support this (@malfet ?) and it wasn't caught by the CI since the Python versions there come from CentOS as described in [PEP 513](https://www.python.org/dev/peps/pep-0513).

It seems to be an unresolved issue, see https://github.com/pypa/distutils/issues/17, looks like setuptools [reverted](https://github.com/pypa/setuptools/pull/2376) their move away from `distutils` due to that issue. The `SETUPTOOLS_USE_DISTUTILS` fix also may not here since we manually specify the install location in cmake.

**Comment by xkszltl** [comment]
@driazati 
If for backward compatibility you may simply symlink between `python3/dist-packages/torch` and `python3.x/site-packages/torch`.

We've been using cmake+pip build for a long while without issue, so I believe that's supported in reality, no matter official or not.
Requiring conda would be a bad idea, because its own environment won't be compatible with other things on system.
IMHO conda is kind of "evil", in the sense of env contamination.

**Comment by xkszltl** [comment]
On Ubuntu 22.04, saw it moved to python3.10/dist-packages recently.

**Pull request closed without merging**